src/fsharp/FSharp.Core/mailbox.fs

```diff
@@ -350,7 +355,7 @@ namespace Microsoft.FSharp.Control
         member _.UnsafeMessageQueueContents = mailbox.UnsafeContents
 #endif
 
-        member x.Start() =
+        member private x.StartWith(startAsync: AsyncStarter.Func) =
```

Comment:
This should just be called `PrepareStart` and return `p`

        member private x.PrepareStart(startAsync: AsyncStarter.Func) = ...
            if started then
                raise (new InvalidOperationException(SR.GetString(SR.mailboxProcessorAlreadyStarted)))
            else
                started <- true
                // Protect the execution and send errors to the event.
                // Note that exception stack traces are lost in this design - in an extended design
                // the event could propagate an ExceptionDispatchInfo instead of an Exception.
                async { try
                            do! body x
                        with exn ->
                            errorEvent.Trigger exn }

        member x.StartImmediate() =
            let p = x.PrepareStart()
            Async.StartImmediate(computation=p, cancellationToken=ct) 

        member x.Start() =
            let p = x.PrepareStart()
            Async.Start(computation=p, cancellationToken=ct)

Reply:
Done